nodebuilder/share/types/get_range_result.go

```diff
@@ -0,0 +1,126 @@
+package types
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/cometbft/cometbft/crypto/merkle"
+	tmbytes "github.com/cometbft/cometbft/libs/bytes"
+	tmproto "github.com/cometbft/cometbft/proto/tendermint/types"
+	"github.com/cometbft/cometbft/types"
+
+	"github.com/celestiaorg/celestia-app/v4/pkg/da"
+	libshare "github.com/celestiaorg/go-square/v2/share"
+	"github.com/celestiaorg/nmt"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/shwap"
+)
+
+// GetRangeResult wraps the return value of the GetRange endpoint
+// because Json-RPC doesn't support more than two return values.
+type GetRangeResult struct {
+	Shares []libshare.Share
+	Proof  *types.ShareProof
+}
+
+func NewGetRangeResult(
+	startIndex, endIndex int,
+	rngdata *shwap.RangeNamespaceData,
+	dah *da.DataAvailabilityHeader,
+) (*GetRangeResult, error) {
+	ns, err := shwap.ParseNamespace(rngdata.Shares, startIndex, endIndex)
+	if err != nil {
+		return nil, err
+	}
+
+	odsSize := len(dah.RowRoots) / 2
+	startRow := startIndex / odsSize
+	endRow := (endIndex - 1) / odsSize
+
+	nmtProofs := make([]*nmt.Proof, len(rngdata.Shares))
+	nmtProofs[0] = rngdata.FirstIncompleteRowProof
+	if startRow != endRow {
+		nmtProofs[len(rngdata.Shares)-1] = rngdata.LastIncompleteRowProof
+	}
+	for i, nmtProof := range nmtProofs {
+		if nmtProof != nil {
+			continue
+		}
+		extendedShares, err := share.ExtendShares(rngdata.Shares[i])
+		if err != nil {
+			return nil, err
+		}
+		proof, err := shwap.GenerateSharesProofs(
+			startRow+i,
+			0,
+			odsSize,
+			odsSize,
+			extendedShares,
+		)
+		if err != nil {
+			return nil, err
+		}
+		nmtProofs[i] = proof
+	}
+	coreProofs := toCoreNMTProof(nmtProofs)
+
+	root, allProofs := merkle.ProofsFromByteSlices(append(dah.RowRoots, dah.ColumnRoots...))
+	if !bytes.Equal(root, dah.Hash()) {
```

Comment:
this check feels unnecessary as both root and dah.Hash are based on dah. only seems to be validating the correctness of ProofsFromByteSlices everytime?

Reply:
Just a sanity check 🙂

Reply:
Let's remove then.